In [1]:
# importing libraries:
import pandas as pd
import numpy as np


In [2]:
# reading the csv file
df = pd.read_csv(r"D:\SDBI\4th Semester\Mashine Learning\Datasets\credit.csv")

In [3]:
# looking at the data
df.head()

checking_balance  months_loan_duration credit_history               purpose  \
0           < 0 DM                     6       critical  furniture/appliances   
1       1 - 200 DM                    48           good  furniture/appliances   
2          unknown                    12       critical             education   
3           < 0 DM                    42           good  furniture/appliances   
4           < 0 DM                    24           poor                   car   

   amount savings_balance employment_duration  percent_of_income  \
0    1169         unknown           > 7 years                  4   
1    5951        < 100 DM         1 - 4 years                  2   
2    2096        < 100 DM         4 - 7 years                  2   
3    7882        < 100 DM         4 - 7 years                  2   
4    4870        < 100 DM         1 - 4 years                  3   

   years_at_residence  age other_credit housing  existing_loans_count  \
0                   4   67         none     own                     2   
1                   2   22         none     own                     1   
2                   3   49         none     own                     1   
3                   4   45         none   other                     1   
4                   4   53         none   other                     2   

         job  dependents phone default  
0    skilled           1   yes      no  
1    skilled           1    no     yes  
2  unskilled           2    no      no  
3    skilled           2    no      no  
4    skilled           2    no     yes

In [4]:
df.tail()

checking_balance  months_loan_duration credit_history  \
995          unknown                    12           good   
996           < 0 DM                    30           good   
997          unknown                    12           good   
998           < 0 DM                    45           good   
999       1 - 200 DM                    45       critical   

                  purpose  amount savings_balance employment_duration  \
995  furniture/appliances    1736        < 100 DM         4 - 7 years   
996                   car    3857        < 100 DM         1 - 4 years   
997  furniture/appliances     804        < 100 DM           > 7 years   
998  furniture/appliances    1845        < 100 DM         1 - 4 years   
999                   car    4576    100 - 500 DM          unemployed   

     percent_of_income  years_at_residence  age other_credit housing  \
995                  3                   4   31         none     own   
996                  4                   4   40         none     own   
997                  4                   4   38         none     own   
998                  4                   4   23         none   other   
999                  3                   4   27         none     own   

     existing_loans_count         job  dependents phone default  
995                     1   unskilled           1    no      no  
996                     1  management           1   yes      no  
997                     1     skilled           1    no      no  
998                     1     skilled           1   yes     yes  
999                     1     skilled           1    no      no

In [5]:
df.sample(5)

checking_balance  months_loan_duration credit_history  \
624           < 0 DM                    18           good   
613           < 0 DM                    24      very good   
787          unknown                    48       critical   
442       1 - 200 DM                    20           poor   
447       1 - 200 DM                     7           good   

                  purpose  amount savings_balance employment_duration  \
624  furniture/appliances    2600        < 100 DM         1 - 4 years   
613                   car    3632        < 100 DM         1 - 4 years   
787                   car    2751         unknown           > 7 years   
442                  car0    2629        < 100 DM         1 - 4 years   
447  furniture/appliances    2576        < 100 DM         1 - 4 years   

     percent_of_income  years_at_residence  age other_credit housing  \
624                  4                   4   65         none   other   
613                  1                   4   22         bank    rent   
787                  4                   3   38         none     own   
442                  2                   3   29         bank     own   
447                  2                   2   35         none     own   

     existing_loans_count      job  dependents phone default  
624                     2  skilled           1    no     yes  
613                     1  skilled           1    no      no  
787                     2  skilled           2   yes      no  
442                     2  skilled           1   yes      no  
447                     1  skilled           1    no      no

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 17 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   checking_balance      1000 non-null   object
 1   months_loan_duration  1000 non-null   int64 
 2   credit_history        1000 non-null   object
 3   purpose               1000 non-null   object
 4   amount                1000 non-null   int64 
 5   savings_balance       1000 non-null   object
 6   employment_duration   1000 non-null   object
 7   percent_of_income     1000 non-null   int64 
 8   years_at_residence    1000 non-null   int64 
 9   age                   1000 non-null   int64 
 10  other_credit          1000 non-null   object
 11  housing               1000 non-null   object
 12  existing_loans_count  1000 non-null   int64 
 13  job                   1000 non-null   object
 14  dependents            1000 non-null   int64 
 15  phone                 1000 non-null   o

**Inference**:

- The data doesn't have missing values although it has NaN values termed as unknown
- The data has 1000 entries and 17 variables
- There are 7 integer data type and 10 object data type

In [7]:
df.describe()

months_loan_duration        amount  percent_of_income  \
count           1000.000000   1000.000000        1000.000000   
mean              20.903000   3271.258000           2.973000   
std               12.058814   2822.736876           1.118715   
min                4.000000    250.000000           1.000000   
25%               12.000000   1365.500000           2.000000   
50%               18.000000   2319.500000           3.000000   
75%               24.000000   3972.250000           4.000000   
max               72.000000  18424.000000           4.000000   

       years_at_residence          age  existing_loans_count   dependents  
count         1000.000000  1000.000000           1000.000000  1000.000000  
mean             2.845000    35.546000              1.407000     1.155000  
std              1.103718    11.375469              0.577654     0.362086  
min              1.000000    19.000000              1.000000     1.000000  
25%              2.000000    27.000000              1.000000     1.000000  
50%              3.000000    33.000000              1.000000     1.000000  
75%              4.000000    42.000000              2.000000     1.000000  
max              4.000000    75.000000              4.000000     2.000000

**Inference** -

- The duration of loans is minimum of 4 months and maximum of 72 months with average of 20 months.
- The minimum loan amount taken is 250 dollars and maximum of 18424 dollars.
- The monthly installment's percentage from a customer's income  range from min 1% and max of 4% and average 2% - 3%.
- The minimum loan holder age is 19 with average of 35 years and maximum is 75 years of age.
- The minimum loans a customer has is 1 and and max of 4.
- The no of dependents on the customer is min of 1 and max of 2.
- The number of customer's having owning a house is min of 0 and max of 1.
- There are 0.02 percent of customers who are unemployed.

### Converting miss-leading data types

In [8]:
# Calling object var in a list

col = df.loc[:,df.dtypes == 'object'].columns.tolist()
df.loc[:,df.dtypes == 'object'].head()

checking_balance credit_history               purpose savings_balance  \
0           < 0 DM       critical  furniture/appliances         unknown   
1       1 - 200 DM           good  furniture/appliances        < 100 DM   
2          unknown       critical             education        < 100 DM   
3           < 0 DM           good  furniture/appliances        < 100 DM   
4           < 0 DM           poor                   car        < 100 DM   

  employment_duration other_credit housing        job phone default  
0           > 7 years         none     own    skilled   yes      no  
1         1 - 4 years         none     own    skilled    no     yes  
2         4 - 7 years         none     own  unskilled    no      no  
3         4 - 7 years         none   other    skilled    no      no  
4         1 - 4 years         none   other    skilled    no     yes

In [9]:
# changing the dtype of object 
print(df.shape)
for i in df.columns:
    if df[i].dtype == 'object':
        df[i] = pd.Categorical(df[i])
print(df.shape)

(1000, 17)
(1000, 17)


In [10]:
for i in df.columns.tolist():
    print("---"*15)
    print(i)
    print("---"*5)
    print(df[i].value_counts())

---------------------------------------------
checking_balance
---------------
unknown       394
< 0 DM        274
1 - 200 DM    269
> 200 DM       63
Name: checking_balance, dtype: int64
---------------------------------------------
months_loan_duration
---------------
24    184
12    179
18    113
36     83
6      75
15     64
9      49
48     48
30     40
21     30
10     28
60     13
27     13
42     11
11      9
20      8
8       7
4       6
45      5
7       5
39      5
14      4
13      4
33      3
28      3
54      2
16      2
22      2
47      1
5       1
26      1
72      1
40      1
Name: months_loan_duration, dtype: int64
---------------------------------------------
credit_history
---------------
good         530
critical     293
poor          88
very good     49
perfect       40
Name: credit_history, dtype: int64
---------------------------------------------
purpose
---------------
furniture/appliances    473
car                     337
business                 97
educati

In [11]:
d = {}
for i in ["checking_balance","credit_history","savings_balance","employment_duration","phone","default"]:
    d[i] = df[i].value_counts().to_dict()

In [12]:
d

{'checking_balance': {'unknown': 394,
  '< 0 DM': 274,
  '1 - 200 DM': 269,
  '> 200 DM': 63},
 'credit_history': {'good': 530,
  'critical': 293,
  'poor': 88,
  'very good': 49,
  'perfect': 40},
 'savings_balance': {'< 100 DM': 603,
  'unknown': 183,
  '100 - 500 DM': 103,
  '500 - 1000 DM': 63,
  '> 1000 DM': 48},
 'employment_duration': {'1 - 4 years': 339,
  '> 7 years': 253,
  '4 - 7 years': 174,
  '< 1 year': 172,
  'unemployed': 62},
 'phone': {'no': 596, 'yes': 404},
 'default': {'no': 700, 'yes': 300}}

In [13]:
pp = {
    "checking_balance" : {"unknown":0,"< 0 DM":1, "1 - 200 DM":2,"> 200 DM":3},
    "credit_history" : {"critical": 0,"poor":1,"good":2,"very good":3,"perfect":4},
    "savings_balance" : {"unknown":0,"< 100 DM":1, "100 - 500 DM":2,"500 - 1000 DM":3,"> 1000 DM":4},
    "employment_duration":{"unemployed":0,"< 1 year":1,"1 - 4 years":2,"> 7 years":4,"4 - 7 years":3,},
    "phone":{"no":0,"yes":1},
    "default":{"no":0,"yes":1}
    }
obj = ["purpose","housing","other_credit","job"]

In [14]:

df.select_dtypes(include = 'object')

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, ...]

[1000 rows x 0 columns]

In [15]:
df = df.replace(pp)
df = pd.get_dummies(df,columns =obj)

In [16]:
df.head()

checking_balance  months_loan_duration credit_history  amount  \
0                1                     6              0    1169   
1                2                    48              2    5951   
2                0                    12              0    2096   
3                1                    42              2    7882   
4                1                    24              1    4870   

  savings_balance employment_duration  percent_of_income  years_at_residence  \
0               0                   4                  4                   4   
1               1                   2                  2                   2   
2               1                   3                  2                   3   
3               1                   3                  2                   4   
4               1                   2                  3                   4   

   age  existing_loans_count  ...  housing_other housing_own housing_rent  \
0   67                     2  ...              0           1            0   
1   22                     1  ...              0           1            0   
2   49                     1  ...              0           1            0   
3   45                     1  ...              1           0            0   
4   53                     2  ...              1           0            0   

   other_credit_bank  other_credit_none  other_credit_store  job_management  \
0                  0                  1                   0               0   
1                  0                  1                   0               0   
2                  0                  1                   0               0   
3                  0                  1                   0               0   
4                  0                  1                   0               0   

   job_skilled  job_unemployed  job_unskilled  
0            1               0              0  
1            1               0              0  
2            0               0              1  
3            1               0              0  
4            1               0              0  

[5 rows x 29 columns]

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 29 columns):
 #   Column                        Non-Null Count  Dtype   
---  ------                        --------------  -----   
 0   checking_balance              1000 non-null   category
 1   months_loan_duration          1000 non-null   int64   
 2   credit_history                1000 non-null   category
 3   amount                        1000 non-null   int64   
 4   savings_balance               1000 non-null   category
 5   employment_duration           1000 non-null   category
 6   percent_of_income             1000 non-null   int64   
 7   years_at_residence            1000 non-null   int64   
 8   age                           1000 non-null   int64   
 9   existing_loans_count          1000 non-null   int64   
 10  dependents                    1000 non-null   int64   
 11  phone                         1000 non-null   category
 12  default                       1000 non-null   cat

In [18]:
x = df.drop(columns = 'default',axis = 1)
y = df["default"]

In [19]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(x,y,test_size = 0.30,random_state = 23) 

In [20]:
from sklearn.tree import DecisionTreeClassifier

dt_model = DecisionTreeClassifier(criterion ='gini',random_state = 23)
dt_model.fit(X_train,y_train)

DecisionTreeClassifier(random_state=23)

In [21]:
print(dt_model.score(X_train,y_train))
print(dt_model.score(X_test,y_test))

1.0
0.6933333333333334


In [22]:
import warnings
warnings.filterwarnings('ignore')

In [23]:
from sklearn.metrics import roc_auc_score

In [24]:
md = [3,5,10,25]
ml = [15,30 ,70]
ms = [15,30,70]
model_performance_data = []

for i in md:
    for j in ml:
        for k in ms:
            
            mp = DecisionTreeClassifier(criterion = 'gini',max_depth = i, min_samples_leaf = j,random_state = 23)
            dt_model = mp.fit(X_train,y_train)
            test_pred = dt_model.predict_proba(X_test)
            train_pred = dt_model.predict_proba(X_train)
            
            t1 = "max_depth" + str(i)+"_min_leaf"+str(j)+"_min_spit"+str(k)
            t2 = [t1,round(roc_auc_score(y_train,train_pred[:,1]),4),round(roc_auc_score(y_test,test_pred[:,1]),4)]
            
            model_performance_data.append(t2)

# Post Puring

In [25]:
apl = [0.001,0.005,0.01,0.05,0.1]
for k in apl:
    
    mp =DecisionTreeClassifier(criterion ='gini',ccp_alpha = k,random_state = 23)
    dt_model = mp.fit(X_train,y_train)
    test_pred = dt_model.predict_proba(X_test)
    train_pred = dt_model.predict_proba(X_train)
    
    t1 = "ccp_alpha"+str(k)
    t2 = [t1,round(roc_auc_score(y_train,train_pred[:,1]),4),round(roc_auc_score(y_test,test_pred[:,1]),4)]
    
    model_performance_data.append(t2)

In [26]:
model_performance_df = pd.DataFrame(model_performance_data)
model_performance_df.rename(columns ={0 :'DESC',1:'Train_AUC',2:'Test_AUC'},inplace = True)

In [27]:
model_performance_df.sort_values(by = ['Test_AUC','Train_AUC'],ascending = [False,False])

DESC  Train_AUC  Test_AUC
18  max_depth10_min_leaf15_min_spit15     0.8601    0.7523
19  max_depth10_min_leaf15_min_spit30     0.8601    0.7523
20  max_depth10_min_leaf15_min_spit70     0.8601    0.7523
27  max_depth25_min_leaf15_min_spit15     0.8601    0.7523
28  max_depth25_min_leaf15_min_spit30     0.8601    0.7523
29  max_depth25_min_leaf15_min_spit70     0.8601    0.7523
9    max_depth5_min_leaf15_min_spit15     0.8284    0.7503
10   max_depth5_min_leaf15_min_spit30     0.8284    0.7503
11   max_depth5_min_leaf15_min_spit70     0.8284    0.7503
37                     ccp_alpha0.005     0.7787    0.7356
0    max_depth3_min_leaf15_min_spit15     0.7778    0.7197
1    max_depth3_min_leaf15_min_spit30     0.7778    0.7197
2    max_depth3_min_leaf15_min_spit70     0.7778    0.7197
15   max_depth5_min_leaf70_min_spit15     0.7611    0.7114
16   max_depth5_min_leaf70_min_spit30     0.7611    0.7114
17   max_depth5_min_leaf70_min_spit70     0.7611    0.7114
24  max_depth10_min_leaf70_min_spit15     0.7611    0.7114
25  max_depth10_min_leaf70_min_spit30     0.7611    0.7114
26  max_depth10_min_leaf70_min_spit70     0.7611    0.7114
33  max_depth25_min_leaf70_min_spit15     0.7611    0.7114
34  max_depth25_min_leaf70_min_spit30     0.7611    0.7114
35  max_depth25_min_leaf70_min_spit70     0.7611    0.7114
12   max_depth5_min_leaf30_min_spit15     0.8090    0.7090
13   max_depth5_min_leaf30_min_spit30     0.8090    0.7090
14   max_depth5_min_leaf30_min_spit70     0.8090    0.7090
6    max_depth3_min_leaf70_min_spit15     0.7598    0.7069
7    max_depth3_min_leaf70_min_spit30     0.7598    0.7069
8    max_depth3_min_leaf70_min_spit70     0.7598    0.7069
3    max_depth3_min_leaf30_min_spit15     0.7642    0.7031
4    max_depth3_min_leaf30_min_spit30     0.7642    0.7031
5    max_depth3_min_leaf30_min_spit70     0.7642    0.7031
38                      ccp_alpha0.01     0.7225    0.7001
21  max_depth10_min_leaf30_min_spit15     0.8148    0.6896
22  max_depth10_min_leaf30_min_spit30     0.8148    0.6896
23  max_depth10_min_leaf30_min_spit70     0.8148    0.6896
30  max_depth25_min_leaf30_min_spit15     0.8148    0.6896
31  max_depth25_min_leaf30_min_spit30     0.8148    0.6896
32  max_depth25_min_leaf30_min_spit70     0.8148    0.6896
36                     ccp_alpha0.001     0.9999    0.6170
39                      ccp_alpha0.05     0.5000    0.5000
40                       ccp_alpha0.1     0.5000    0.5000

# Second round of Tunning

In [28]:
md = [5,7,9]
ml = [15,20,25]
apl = [0.003,0.004,0.005,0.006,0.007]

model_performance_data =[]

for i in md:
    for j in ml:
        for k in apl:
            
            mp = DecisionTreeClassifier(criterion = 'gini',max_depth = i, min_samples_leaf = j,random_state = 23)
            dt_model = mp.fit(X_train,y_train)
            test_pred = dt_model.predict_proba(X_test)
            train_pred = dt_model.predict_proba(X_train)
            
            t1 = "max_depth" + str(i)+"_min_leaf"+str(j)+"_min_spit"+str(k)
            t2 = [t1,round(roc_auc_score(y_train,train_pred[:,1]),4),round(roc_auc_score(y_test,test_pred[:,1]),4)]
            
            model_performance_data.append(t2)

In [29]:
model_performance_df = pd.DataFrame(model_performance_data)
model_performance_df.rename(columns ={0 :'DESC',1:'Train_AUC',2:'Test_AUC'},inplace = True)

In [30]:
model_performance_df.sort_values(by = ['Test_AUC','Train_AUC'],ascending = [False,False])

DESC  Train_AUC  Test_AUC
20  max_depth7_min_leaf20_min_spit0.003     0.8503    0.7568
21  max_depth7_min_leaf20_min_spit0.004     0.8503    0.7568
22  max_depth7_min_leaf20_min_spit0.005     0.8503    0.7568
23  max_depth7_min_leaf20_min_spit0.006     0.8503    0.7568
24  max_depth7_min_leaf20_min_spit0.007     0.8503    0.7568
35  max_depth9_min_leaf20_min_spit0.003     0.8503    0.7568
36  max_depth9_min_leaf20_min_spit0.004     0.8503    0.7568
37  max_depth9_min_leaf20_min_spit0.005     0.8503    0.7568
38  max_depth9_min_leaf20_min_spit0.006     0.8503    0.7568
39  max_depth9_min_leaf20_min_spit0.007     0.8503    0.7568
15  max_depth7_min_leaf15_min_spit0.003     0.8581    0.7559
16  max_depth7_min_leaf15_min_spit0.004     0.8581    0.7559
17  max_depth7_min_leaf15_min_spit0.005     0.8581    0.7559
18  max_depth7_min_leaf15_min_spit0.006     0.8581    0.7559
19  max_depth7_min_leaf15_min_spit0.007     0.8581    0.7559
30  max_depth9_min_leaf15_min_spit0.003     0.8601    0.7523
31  max_depth9_min_leaf15_min_spit0.004     0.8601    0.7523
32  max_depth9_min_leaf15_min_spit0.005     0.8601    0.7523
33  max_depth9_min_leaf15_min_spit0.006     0.8601    0.7523
34  max_depth9_min_leaf15_min_spit0.007     0.8601    0.7523
5   max_depth5_min_leaf20_min_spit0.003     0.8272    0.7509
6   max_depth5_min_leaf20_min_spit0.004     0.8272    0.7509
7   max_depth5_min_leaf20_min_spit0.005     0.8272    0.7509
8   max_depth5_min_leaf20_min_spit0.006     0.8272    0.7509
9   max_depth5_min_leaf20_min_spit0.007     0.8272    0.7509
0   max_depth5_min_leaf15_min_spit0.003     0.8284    0.7503
1   max_depth5_min_leaf15_min_spit0.004     0.8284    0.7503
2   max_depth5_min_leaf15_min_spit0.005     0.8284    0.7503
3   max_depth5_min_leaf15_min_spit0.006     0.8284    0.7503
4   max_depth5_min_leaf15_min_spit0.007     0.8284    0.7503
25  max_depth7_min_leaf25_min_spit0.003     0.8168    0.7381
26  max_depth7_min_leaf25_min_spit0.004     0.8168    0.7381
27  max_depth7_min_leaf25_min_spit0.005     0.8168    0.7381
28  max_depth7_min_leaf25_min_spit0.006     0.8168    0.7381
29  max_depth7_min_leaf25_min_spit0.007     0.8168    0.7381
40  max_depth9_min_leaf25_min_spit0.003     0.8168    0.7381
41  max_depth9_min_leaf25_min_spit0.004     0.8168    0.7381
42  max_depth9_min_leaf25_min_spit0.005     0.8168    0.7381
43  max_depth9_min_leaf25_min_spit0.006     0.8168    0.7381
44  max_depth9_min_leaf25_min_spit0.007     0.8168    0.7381
10  max_depth5_min_leaf25_min_spit0.003     0.8105    0.7267
11  max_depth5_min_leaf25_min_spit0.004     0.8105    0.7267
12  max_depth5_min_leaf25_min_spit0.005     0.8105    0.7267
13  max_depth5_min_leaf25_min_spit0.006     0.8105    0.7267
14  max_depth5_min_leaf25_min_spit0.007     0.8105    0.7267

# Grid Search

In [31]:
from sklearn.model_selection import GridSearchCV

grid = {'ccp_alpha' : [0.002, 0.003, 0.004],
           'max_depth' : [7,8,9],
           'min_samples_leaf':[15,18,20,22,25]
       }

dt_model = DecisionTreeClassifier(random_state = 23)
grid_search = GridSearchCV(estimator = dt_model, param_grid = grid,cv = 5, verbose = True)
grid_search.fit(X_train,y_train)

Fitting 5 folds for each of 45 candidates, totalling 225 fits


GridSearchCV(cv=5, estimator=DecisionTreeClassifier(random_state=23),
             param_grid={'ccp_alpha': [0.002, 0.003, 0.004],
                         'max_depth': [7, 8, 9],
                         'min_samples_leaf': [15, 18, 20, 22, 25]},
             verbose=True)

In [32]:
grid_search.best_estimator_

DecisionTreeClassifier(ccp_alpha=0.003, max_depth=7, min_samples_leaf=18,
                       random_state=23)

# Final Model

In [33]:
mp = DecisionTreeClassifier(criterion = 'gini',max_depth = 7, min_samples_leaf = 18, ccp_alpha = 0.003,random_state = 23)
dt_model = mp.fit(X_train,y_train)

test_pred = dt_model.predict_proba(X_test)
train_pred = dt_model.predict_proba(X_train)

print("Performance on training data :",round(roc_auc_score(y_train,train_pred[:,1]),4))
print("performance on testing data :",round(roc_auc_score(y_test,test_pred[:,1]),4))

Performance on training data : 0.8102
performance on testing data : 0.7453
